### PR TITLE
Block Editor: Pressing enter at start of the text block prepends a block

### PR DESCRIPTION
**Pull request by gziolo**
### Before

Previously, you would have to use `Insert before` feature to prepend a block. Using the keyboard to achieve the same goal with the Paragraph and Heading blocks was only possible. You could do it by pressing `enter` with the caret at the start of the text block that contains content.

https://user-images.githubusercontent.com/699132/179499983-71784389-fd2d-4c5f-a32d-0fdf9f7b1ac5.mov

### After

This PR proposes bringing the same functionality that exists for the Paragraph and Heading blocks to other text blocks:

- Code
- Preformatted
- Read More
- Site Tagline
- Site Title
- Verse

https://user-images.githubusercontent.com/699132/179212805-345b4077-8c45-46e8-a075-7312f0690fe7.mov

It might be good to check if we can add it as a follow-up to other blocks like:

- Quote – the most recent version uses nested blocks, so it's a more complex task
- Pullquote – it's similar to the old Quote block, but the question is whether this block shouldn't follow the refactoring done to use nested blocks

I'm keeping this PR scoped because the aforementioned blocks have more complex usage or `RichText` component.

Another potential follow-up task is to improve the behavior when entering multiple new lines at the end of the block. Currently, you also get trapped inside text blocks that allow new lines in the content presented in the "before" video.

## Why?

To improve the writing flow experience and have more natural and consistent interaction when pressing enter when the caret is placed at the start of the text blocks.

## How?

The initial proposal mirrors the existing unstable method `__unstableOnSplitAtEnd` that inserts the block when the `enter` key is pressed at the end of the block. This PR proposes adding `__unstableOnSplitAtStart` function callback to `RichText` that gets called in the following conditions:
- there is any content in the contenteditable controlled by the `RichText` component
- the caret is at the beginning of the block

### Technical considerations

I believe this could get also promoted to stable API with the name `onEnterAtStart` together with `onEnterAtEnd` as it is something that proved to be useful with `__unstableOnSplitAtEnd`.

Optionally, we might also consider making both `onEnterAtStart` and `onEnterAtEnd` more abstracted in `RichText` from `@wordpress/block-editor` and allow passing `true`, example:

```jsx
<RichText onEnterAtStart onEnterAtEnd />
```

that would default to the callback that automatically finds the current client id of the block and inserts a new default block before/after the block.

## Testing Instructions

1. Open the site editor.
2. Add a new empty template or remove all blocks from the currently edited template.
3. Add one of the listed blocks, e.g. Code block.
4. Type something, so it contains any content.
5. Move the caret to the start of the block and press `enter`.
6. New default block should get prepended.

### CODE REVIEW

File: packages/block-library/src/verse/edit.js [hunk 1]
```diff
@@ -53,6 +57,9 @@ export default function VerseEdit( {
 				placeholder={ __( 'Write verse…' ) }
 				onRemove={ onRemove }
 				onMerge={ mergeBlocks }
+				__unstableOnSplitAtStart={ () => {
+					insertBeforeBlock( clientId );
```
Comment (ntsekouras) [inline packages/block-library/src/verse/edit.js:61]: Do you think we should create a similar function to [onInsertBlocksAfter](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-list/block.js#L287) ?

Comment (gziolo) [reply packages/block-library/src/verse/edit.js:61]: What benefits do you see with introducing another prop passed a few levels down?

Comment (ntsekouras) [reply packages/block-library/src/verse/edit.js:61]: Not much to be honest, but the minor things could be:
1. consistency with the `after`
2. more flexibility to what block/blocks to insert - although I don't have a use case on top of my mind that we would want something different. Maybe that might be useful to blocks with innerBlocks that allow specific blocks similar to new List? I haven't checked in depth though.. 

No strong opinions here 😄

Comment (gziolo) [reply packages/block-library/src/verse/edit.js:61]: `onInsertBlocksAfter` / `insertBlocksAfter` was added before React hooks were used. That's why it was exposed through `withDispatch` HOC. I would personally go in the opposite direction and refactor code in other places to use `useDispatch` and pick the right action for the job: `insertBlocks` is a general one that gives maximum flexibility, `insertBeforeBlock` is great when you want to prepend the default block. There is also `insertAfterBlock` available. If necessary, we can also extend those more specific actions to accept a different block(s) optionally.

Comment (gziolo) [reply packages/block-library/src/verse/edit.js:61]: See also my note in the description:

> Optionally, we might also consider making both `onEnterAtStart` and `onEnterAtEnd` more abstracted in RichText from `@wordpress/block-editor` and allow passing `true`, example:
>
> `<RichText onEnterAtStart onEnterAtEnd />`
>
> that would default to the callback that automatically finds the current client id of the block and inserts a new default block before/after the block.

Comment (ntsekouras) [reply packages/block-library/src/verse/edit.js:61]: Sounds good!

**Review by ellatrix — CHANGES_REQUESTED** [review]
Looking at the video, focus moves to the newly inserted block before the current block. That seems wrong. If you press Enter at the start of a heading for example, you'll notice that focus stays in the heading.

Secondly, I'm not sure if this behaviour makes sense to me. Splitting from the start makes sense for blocks that don't insert a newline on Enter. But code and verse etc. do insert newlines on Enter. So why should it split when at the start of the block? It's now impossible to create a new line before the existing content and add something there.

It could however make sense to add split on **double** Enter for these blocks.

**Comment by gziolo** [comment]
@ellatrix, thank you for feedback.

> It could however make sense to add split on double Enter for these blocks.

This is how pressing `enter` key at the start of the Verse block behaves in `trunk`:

https://user-images.githubusercontent.com/699132/179963675-5f3e3a3d-be32-4190-abe3-09bc0fbc8315.mov

It's hard to tell whether double `enter` would be an optimal choice.

I see that in the Classic block you can use arrow up/down to insert a paragraph after/before. Maybe that could be a way to improve the current writing flow?

https://user-images.githubusercontent.com/699132/179964249-e4372bb6-9bfa-4105-876d-71809c343c15.mov

**Comment by ellatrix** [comment]
> It's hard to tell whether double enter would be an optimal choice.

Yes, it only really makes sense for inserting it after the block, not before.

> I see that in the Classic block you can use arrow up/down to insert a paragraph after/before. Maybe that could be a way to improve the current writing flow?

That sounds interesting, although it's hard with FSE and even in the post editor where you could want to navigate to the title.
This actually reminds me of something I wanted to do a while back: it would be cool if arrow key navigation to other blocks had an intermediary step that focusses the in-between inserter. So that could solve your issue here. On arrow up, focus moves to the inserter before the block before moving to the next block so you can insert something. I have a branch somewhere locally that attempts this.

**Comment by gziolo** [comment]
> it would be cool if arrow key navigation to other blocks had an intermediary step that focusses the in-between inserter. 

In the case where you interact with the first block, then there is no in-between inserter, but maybe there should be something like that.

> On arrow up, focus moves to the inserter before the block before moving to the next block so you can insert something. I have a branch somewhere locally that attempts this.

It would be great to try that instead then 👍🏻

**Comment by mtias** [comment]
I feel focusing on in between state would make the experience of moving through text worse or less intuitive

**Comment by gziolo** [comment]
I didn't make any further progress on this PR, and I won't have time in the near future. I'm closing this PR but everyone is invited to take it over.

**Pull request closed without merging**